File: Day27/Part2/equal_souls.py
from typing import List, Tuple
import sys
sys.setrecursionlimit(1 << 20)

def min_diff(n: int, edges: List[Tuple[int,int]], power: List[int]) -> int:
    g = [[] for _ in range(n)]
    for u,v in edges:
        u-=1; v-=1
        g[u].append(v); g[v].append(u)
    S = sum(power)
    sub = [0]*n
    seen = [False]*n
    ans = float('inf')
    def dfs(u: int, p: int):
        nonlocal ans
        seen[u] = True
        s = power[u]
        for v in g[u]:
            if v==p: continue
            dfs(v,u)
            s += sub[v]
            ans = min(ans, abs(S - 2*sub[v]))
        sub[u] = s
    dfs(0,-1)
    return int(ans)

if __name__ == "__main__":
    n = 5
    edges = [(1,2),(1,3),(3,4),(3,5)]
    power = [1,2,3,6,2]
    print(min_diff(n, edges, power))  # Expected 2

File: learnLeetCode_remote/Day27/Part2/equal_souls.cpp
#include <bits/stdc++.h>
using namespace std;

long long min_diff(int n, const vector<pair<int,int>>& edges, const vector<long long>& power){
    vector<vector<int>> g(n);
    for(auto [u,v]: edges){ --u; --v; g[u].push_back(v); g[v].push_back(u); }
    long long S = 0; for(auto x: power) S += x;
    vector<long long> sub(n,0);
    long long ans = (1LL<<62);
    function<void(int,int)> dfs = [&](int u,int p){
        long long s = power[u];
        for(int v: g[u]) if(v!=p){
            dfs(v,u);
            s += sub[v];
            ans = min(ans, llabs(S - 2*sub[v]));
        }
        sub[u]=s;
    };
    dfs(0,-1);
    return ans;
}

int main(){
    int n=5; vector<pair<int,int>> edges={{1,2},{1,3},{3,4},{3,5}}; vector<long long> power={1,2,3,6,2};
    cout << min_diff(n, edges, power) << "\n"; // Expected 2
    return 0;
}
